ext/CEGUI/IrrlichtBaWTexture.cpp

```diff
@@ -0,0 +1,167 @@
+{
+    return cachingName;
+}
+
+const CEGUI::Sizef& IrrlichtBaWTexture::getSize() const
+{
+    bool status = gpuTexture.get();
+
+    if (status)
+    {
+        auto& params = gpuTexture->getCreationParameters().image->getCreationParameters();
+        return { params.extent.width, params.extent.height };
+    }
+    else
+    {
+        assert(status);
+        return { 0, 0 };
+    }
+        
+}
+
+const CEGUI::Sizef& IrrlichtBaWTexture::getOriginalDataSize() const
+{
+    bool status = gpuTexture.get();
+
+    if (status)
+    {
+        auto& params = gpuTexture->getCreationParameters().image->getCreationParameters();
+        return { bufferRowLength ? bufferRowLength : params.extent.width, params.extent.height };
+    }
+    else 
+    {
+        assert(status);
+        return { 0, 0 };
+    }
+}
+
+const CEGUI::Vector2f& IrrlichtBaWTexture::getTexelScaling() const
+{
+    bool status = gpuTexture.get();
+
+    if (status)
+    {
+        auto& params = gpuTexture->getCreationParameters().image->getCreationParameters();
+        auto width = bufferRowLength ? bufferRowLength : params.extent.width;
+
+        return
+        {
+            static_cast<float>(width) == 0.f ? 0.f : 1.f / static_cast<float>(width),
+            static_cast<float>(params.extent.height) == 0.f ? 0.f : 1.f / static_cast<float>(params.extent.height)
+        };
+    }
+    else
+    {
+        assert(status);
+        return { 0, 0 };
+    }
+}
+
+void IrrlichtBaWTexture::loadFromFile(const CEGUI::String& filename, const CEGUI::String& resourceGroup)
+{
+    auto path = resourceGroup + filename;
+    auto assetManager = device->getAssetManager();
+
+    auto params = asset::IAssetLoader::SAssetLoadParams();
+    auto bundle = assetManager->getAsset(path, params);
+    auto cpuImage = asset::IAsset::castDown<asset::ICPUImage>(bundle.getContents().first[0]);
+    assert(cpuImage.get(), "something went wrong while casting, empty pointer detected!");
+
+    bufferRowLength = cpuImage->getRegions().begin()->bufferRowLength;
+
+    asset::ICPUImageView::SCreationParams imageViewInfo;
+    imageViewInfo.image = std::move(cpuImage);
+    imageViewInfo.format = imageViewInfo.image->getCreationParameters().format;
+    imageViewInfo.viewType = asset::IImageView<asset::ICPUImage>::ET_2D;
+    imageViewInfo.flags = static_cast<asset::ICPUImageView::E_CREATE_FLAGS>(0u);
+    imageViewInfo.subresourceRange.baseArrayLayer = 0u;
+    imageViewInfo.subresourceRange.baseMipLevel = 0u;
+    imageViewInfo.subresourceRange.layerCount = imageViewInfo.image->getCreationParameters().arrayLayers;
+    imageViewInfo.subresourceRange.levelCount = imageViewInfo.image->getCreationParameters().mipLevels;
+
+    auto cpuImageView = asset::ICPUImageView::create(std::move(imageViewInfo));
+
+    // cachingName = ""     TODO Okay, so we have to get likely GL id or something similar. Should I cast it to OpenGL image or maybe we should get CPU cache string?
```

Comment:
what does CEGUI use `cachingName` for ?

Reply:
they use for caching afaik, doing like folllowing

```C++
//! container type used to hold Textures we create.
    typedef std::map<String, OpenGLTexture*, StringFastLessCompare
                     CEGUI_MAP_ALLOC(String, OpenGLTexture*)> TextureMap;
    //! Container used to track textures.
    TextureMap d_textures;
```

Reply:
the solution will be https://github.com/buildaworldnet/IrrlichtBAW/pull/480#discussion_r422015730

Reply:
I think its possible to create a "clean" asset manager class (`core::make_smart_refctd_ptr<IAssetManager>()`) that could be used just by the CEGUI renderer components, or at least use an override during loading that could prepend `?CEGUI?` to all the caching paths when inserting into cache and searching the cache